drivers/dht/dht.c

```diff
@@ -94,106 +113,42 @@ static void dht_read_data(gpio_t dev, uint32_t *data, uint8_t *checksum)
-        /* calculate bit length (long 1, short 0) */
-        unsigned long stop = xtimer_now();
-        /* compensate for overflow if needed */
-        if (stop < start) {
-            stop = UINT32_MAX - stop;
-            start = 0;
-        }
-        if ((stop - start) > 40) {
-            /* read 1, set bit */
-            les_bits |= 0x0000000000000001;
-        }
-        else {
-            /* read 0, don't set bit */
-        }
-    }
-
-    *checksum = les_bits & 0x00000000000000FF;
-    *data = (les_bits >> 8) & 0x00000000FFFFFFFF;
-
-    gpio_init(dev, GPIO_DIR_OUT, GPIO_PULLUP);
-    gpio_set(dev);
-}
+    /* read the humidity, temperature, and checksum bits */
+    raw_hum = read(dev->pin, 16);
+    raw_temp = read(dev->pin, 16);
+    csum = (uint8_t)read(dev->pin, 8);
 
-/***********************************************************************
- * public API implementation
- **********************************************************************/
+    /* set pin high again - so we can trigger the next reading by pulling it low
+     * again */
+    gpio_init(dev->pin, GPIO_DIR_OUT, dev->pull);
+    gpio_set(dev->pin);
 
-int dht_init(dht_t *dev, dht_type_t type, gpio_t gpio)
-{
-    DEBUG("dht_init\n");
-
-    dev->gpio = gpio;
-    dev->type = type;
-
-    if (gpio_init(gpio, GPIO_DIR_OUT, GPIO_PULLUP) == -1) {
+    /* validate the checksum */
+    sum = (raw_temp >> 8) + (raw_temp & 0xff) + (raw_hum >> 8) + (raw_hum & 0xff);
+    if ((sum != csum) || (csum == 0)) {
+        DEBUG("error: checksum invalid\n");
         return -1;
     }
-    gpio_set(gpio);
-
-    xtimer_usleep(2000 * MS_IN_USEC);
-
-    DEBUG("dht_init: success\n");
-    return 0;
-}
-
-
-int dht_read_raw(dht_t *dev, dht_data_t *outdata)
-{
-    uint32_t data;
-    uint8_t checksum;
-
-    /* read raw data */
-    dht_read_data(dev->gpio, &data, &checksum);
-
-    /* check checksum */
-    int ret = dht_test_checksum(data, checksum);
-    if (!ret) {
-        DEBUG("checksum fail\n");
-    }
-
-    outdata->humidity = data >> 16;
-    outdata->temperature = data & 0x0000FFFF;
-
-    return (ret - 1); /* take that logic! */
-}
 
-void dht_parse(dht_t *dev, dht_data_t *data, float *outrelhum, float *outtemp)
-{
+    /* parse the RAW values */
+    DEBUG("RAW values: temp: %7i hum: %7i\n", (int)raw_temp, (int)raw_hum);
     switch (dev->type) {
-        case (DHT11):
-            dht_parse_11(data, outrelhum, outtemp);
+        case DHT11:
+            *temp = (int16_t)((raw_temp >> 8) * 10);
+            *hum = (int16_t)((raw_hum >> 8) * 10);
             break;
-
         case DHT22:
-            dht_parse_22(data, outrelhum, outtemp);
-            break;
-
+            *hum = (int16_t)raw_hum;
+            /* if the high-bit is set, the value is negative */
+            if (raw_temp & 0x8000) {
+                *temp = (int16_t)((raw_temp & ~0x8000) * -1);
+            }
+            else {
+                *temp = (int16_t)raw_temp;
+            }
```

Comment:
yes, look at the documentation -> the sensor does return the actual temperature value in
`°C * 10^-1`